CorsixTH/Lua/world.lua

```diff
@@ -1023,6 +1023,16 @@ function World:onTick()
     end
     self.tick_timer = self.tick_rate
 
+    -- Remove entities marked for deletion
+    if self.entity_gc then
+      for i, entity in pairs(self.entities) do
+        if entity.delete then
+          table.remove(self.entities, i)
```

Comment:
Have you verified entities returns to iterative after this process? (so ipairs can be used again)

Reply:
``` lua
local entries = {{delete=true}, {delete=true}, {delet=true}, {delete=true}}
for i, entry in pairs(entries) do
  if entry.delete then
    table.remove(entries, i)
  end
end

local count = 0
for x, y in pairs(entries) do count = count + 1 end
print(count)
```
prints "2" here, which is incorrect.

You simply cannot safely delete entries from a table while iterating over it.

Reply:
It should return `1` since there's a typo `{delet=true}` =]
But yes, looks like even `pairs` isn't safe from this.

Reply:
I see. Option a is to move through entries backwards. 
`local entries = {{delete=true}, {delete=true}, {delete=true}, {delete=true}}
for n = #entries,1,-1 do
  if entries[n].delete then
    table.remove(entries, n)
  end
end

local count = 0
for x, y in pairs(entries) do
    count = count + 1
end
print(count) `

Option b is the function we looked at in discord from https://stackoverflow.com/a/53038524

Reply:
@Alberth289346 what about this?

```lua
local entries = {{delete=true}, {delete=true}, {delete=true}, {delete=true}}
for i=#entries, 1, -1 do
  if entries[i].delete then
    table.remove(entries, i)
    if #entries == 0 then break end
  end
end

local count = 0
for x, y in pairs(entries) do count = count + 1 end
```

left the typo one in so it should return 1 (it did for me), and I think the table remained iterative.

Reply:
Ha, well spotted lewri :)

I agree that backward iteration will work. A nice solution that I didn't realize as a solution yet.

Reply:
Note that a problem only occurs if you have 2 consecutive entries that must be deleted. The second one is never checked for deletion, but if that entry is to be kept then you won't find the problem.